File: app/src/main/java/com/liushu/crazyandroid/ui/main/activity/MainActivity.java
package com.liushu.crazyandroid.ui.main.activity;

import android.animation.AnimatorSet;
import android.animation.ObjectAnimator;
import android.animation.ValueAnimator;
import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.support.v4.app.FragmentTransaction;
import android.view.ViewGroup;
import android.widget.Toast;

import com.flyco.tablayout.CommonTabLayout;
import com.flyco.tablayout.listener.CustomTabEntity;
import com.flyco.tablayout.listener.OnTabSelectListener;
import com.jaydenxiao.common.base.BaseActivity;
import com.jaydenxiao.common.commonutils.LogUtils;
import com.jaydenxiao.common.daynightmodeutils.ChangeModeController;
import com.liushu.crazyandroid.R;
import com.liushu.crazyandroid.app.AppConstant;
import com.liushu.crazyandroid.bean.TabEntity;
import com.liushu.crazyandroid.ui.main.fragment.FirstFragment;
import com.liushu.crazyandroid.ui.main.fragment.FourFragment;
import com.liushu.crazyandroid.ui.main.fragment.SecondFragment;
import com.liushu.crazyandroid.ui.main.fragment.ThirdFragment;

import java.util.ArrayList;

import butterknife.Bind;

public class MainActivity extends BaseActivity {
    @Bind(R.id.tab_layout)
    CommonTabLayout tabLayout;

    private String[] mTitles = {"第一阶段", "第二阶段", "第三阶段", "第四阶段"};
    private int[] mIconUnselectIds = {
            R.mipmap.ic_home_normal, R.mipmap.ic_girl_normal, R.mipmap.ic_video_normal, R.mipmap.ic_care_normal};
    private int[] mIconSelectIds = {
            R.mipmap.ic_home_selected, R.mipmap.ic_girl_selected, R.mipmap.ic_video_selected, R.mipmap.ic_care_selected};
    private ArrayList<CustomTabEntity> mTabEntities = new ArrayList<>();

    private FirstFragment mFirstFragment;
    private SecondFragment mSecondFragment;
    private ThirdFragment mThirdFragment;
    private FourFragment mFourFragment;
    private static int tabLayoutHeight;
    private long exitTime = 0;
    /**
     * 入口
     *
     * @param activity
     */
    public static void startAction(Activity activity) {
        Intent intent = new Intent(activity, MainActivity.class);
        activity.startActivity(intent);
        activity.overridePendingTransition(R.anim.fade_in,
                com.jaydenxiao.common.R.anim.fade_out);
    }

    @Override
    public int getLayoutId() {
        return R.layout.activity_main;
    }

    @Override
    public void initPresenter() {

    }

    @Override
    public void initView() {
        initTab();
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        //切换daynight模式要立即变色的页面
        ChangeModeController.getInstance().init(this, R.attr.class);
        super.onCreate(savedInstanceState);
        //初始化frament
        initFragment(savedInstanceState);
        tabLayout.measure(0, 0);
        tabLayoutHeight = tabLayout.getMeasuredHeight();
    }

    /**
     * 初始化tab
     */
    private void initTab() {
        for (int i = 0; i < mTitles.length; i++) {
            mTabEntities.add(new TabEntity(mTitles[i], mIconSelectIds[i], mIconUnselectIds[i]));
        }
        tabLayout.setTabData(mTabEntities);
        //点击监听
        tabLayout.setOnTabSelectListener(new OnTabSelectListener() {
            @Override
            public void onTabSelect(int position) {
                SwitchTo(position);
            }

            @Override
            public void onTabReselect(int position) {
            }
        });
    }

    /**
     * 初始化碎片
     */
    private void initFragment(Bundle savedInstanceState) {
        FragmentTransaction transaction = getSupportFragmentManager().beginTransaction();
        int currentTabPosition = 0;
        if (savedInstanceState != null) {
            mFirstFragment = (FirstFragment) getSupportFragmentManager().findFragmentByTag("newsMainFragment");
            mSecondFragment = (SecondFragment) getSupportFragmentManager().findFragmentByTag("photosMainFragment");
            mThirdFragment = (ThirdFragment) getSupportFragmentManager().findFragmentByTag("videoMainFragment");
            mFourFragment = (FourFragment) getSupportFragmentManager().findFragmentByTag("careMainFragment");
            currentTabPosition = savedInstanceState.getInt(AppConstant.HOME_CURRENT_TAB_POSITION);
        } else {
            mFirstFragment = new FirstFragment();
            mSecondFragment = new SecondFragment();
            mThirdFragment = new ThirdFragment();
            mFourFragment = new FourFragment();

            transaction.add(R.id.fl_body, mFirstFragment, "newsMainFragment");
            transaction.add(R.id.fl_body, mSecondFragment, "photosMainFragment");
            transaction.add(R.id.fl_body, mThirdFragment, "videoMainFragment");
            transaction.add(R.id.fl_body, mFourFragment, "careMainFragment");
        }
        transaction.commit();
        SwitchTo(currentTabPosition);
        tabLayout.setCurrentTab(currentTabPosition);
    }

    /**
     * 切换
     */
    private void SwitchTo(int position) {
        LogUtils.logd("主页菜单position" + position);
        FragmentTransaction transaction = getSupportFragmentManager().beginTransaction();
        switch (position) {
            //首页
            case 0:
                transaction.hide(mSecondFragment);
                transaction.hide(mThirdFragment);
                transaction.hide(mFourFragment);
                transaction.show(mFirstFragment);
                transaction.commitAllowingStateLoss();
                break;
            //美女
            case 1:
                transaction.hide(mFirstFragment);
                transaction.hide(mThirdFragment);
                transaction.hide(mFourFragment);
                transaction.show(mSecondFragment);
                transaction.commitAllowingStateLoss();
                break;
            //视频
            case 2:
                transaction.hide(mFirstFragment);
                transaction.hide(mSecondFragment);
                transaction.hide(mFourFragment);
                transaction.show(mThirdFragment);
                transaction.commitAllowingStateLoss();
                break;
            //关注
            case 3:
                transaction.hide(mFirstFragment);
                transaction.hide(mSecondFragment);
                transaction.hide(mThirdFragment);
                transaction.show(mFourFragment);
                transaction.commitAllowingStateLoss();
                break;
            default:
                break;
        }
    }

    /**
     * 菜单显示隐藏动画
     *
     * @param showOrHide
     */
    private void startAnimation(boolean showOrHide) {
        final ViewGroup.LayoutParams layoutParams = tabLayout.getLayoutParams();
        ValueAnimator valueAnimator;
        ObjectAnimator alpha;
        if (!showOrHide) {
            valueAnimator = ValueAnimator.ofInt(tabLayoutHeight, 0);
            alpha = ObjectAnimator.ofFloat(tabLayout, "alpha", 1, 0);
        } else {
            valueAnimator = ValueAnimator.ofInt(0, tabLayoutHeight);
            alpha = ObjectAnimator.ofFloat(tabLayout, "alpha", 0, 1);
        }
        valueAnimator.addUpdateListener(new ValueAnimator.AnimatorUpdateListener() {
            @Override
            public void onAnimationUpdate(ValueAnimator valueAnimator) {
                layoutParams.height = (int) valueAnimator.getAnimatedValue();
                tabLayout.setLayoutParams(layoutParams);
            }
        });
        AnimatorSet animatorSet = new AnimatorSet();
        animatorSet.setDuration(500);
        animatorSet.playTogether(valueAnimator, alpha);
        animatorSet.start();
    }

    /**
     * 监听全屏视频时返回键
     */
    @Override
    public void onBackPressed() {
        if (System.currentTimeMillis() - exitTime > 2000) {
            Toast.makeText(getApplicationContext(), "再按一次退出程序", Toast.LENGTH_SHORT).show();
            exitTime = System.currentTimeMillis();
        } else {
            Intent backHome = new Intent(Intent.ACTION_MAIN);
            backHome.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
            backHome.addCategory(Intent.CATEGORY_HOME);
            startActivity(backHome);
        }
    }


    @Override
    protected void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        //奔溃前保存位置
        if (tabLayout != null) {
            outState.putInt(AppConstant.HOME_CURRENT_TAB_POSITION, tabLayout.getCurrentTab());
        }
    }


    @Override
    protected void onDestroy() {
        super.onDestroy();
        ChangeModeController.onDestory();
    }

}

File: app/src/main/java/com/liushu/crazyandroid/ui/main/activity/MainActivity.kt
package com.liushu.crazyandroid.ui.main.activity

import android.animation.AnimatorSet
import android.animation.ObjectAnimator
import android.animation.ValueAnimator
import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.widget.Toast
import com.flyco.tablayout.listener.CustomTabEntity
import com.flyco.tablayout.listener.OnTabSelectListener
import com.jaydenxiao.common.base.BaseActivity
import com.jaydenxiao.common.commonutils.LogUtils
import com.jaydenxiao.common.daynightmodeutils.ChangeModeController
import com.liushu.crazyandroid.R
import com.liushu.crazyandroid.app.AppConstant
import com.liushu.crazyandroid.bean.TabEntity
import com.liushu.crazyandroid.ui.main.fragment.FirstFragment
import com.liushu.crazyandroid.ui.main.fragment.FourFragment
import com.liushu.crazyandroid.ui.main.fragment.SecondFragment
import com.liushu.crazyandroid.ui.main.fragment.ThirdFragment
import kotlinx.android.synthetic.main.activity_main.*
import java.util.*

class MainActivity : BaseActivity() {

    private val mTitles = arrayOf("疯狂android讲义", "android开发艺术探索", "第三阶段", "关于我")
    private val mIconUnselectIds = intArrayOf(R.mipmap.ic_home_normal, R.mipmap.ic_girl_normal, R.mipmap.ic_video_normal, R.mipmap.ic_care_normal)
    private val mIconSelectIds = intArrayOf(R.mipmap.ic_home_selected, R.mipmap.ic_girl_selected, R.mipmap.ic_video_selected, R.mipmap.ic_care_selected)
    private val mTabEntities = ArrayList<CustomTabEntity>()

    private var mFirstFragment: FirstFragment? = null
    private var mSecondFragment: SecondFragment? = null
    private var mThirdFragment: ThirdFragment? = null
    private var mFourFragment: FourFragment? = null
    private var exitTime: Long = 0

    override val layoutId: Int
        get() = R.layout.activity_main


    override fun initView() {
        initTab()
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        //切换daynight模式要立即变色的页面
        ChangeModeController.getInstance().init(this, R.attr::class.java)
        super.onCreate(savedInstanceState)
        //初始化frament
        initFragment(savedInstanceState)
        tab_layout!!.measure(0, 0)
        tabLayoutHeight = tab_layout!!.measuredHeight
    }

    /**
     * 初始化tab
     */
    private fun initTab() {
        for (i in mTitles.indices) {
            mTabEntities.add(TabEntity(mTitles[i], mIconSelectIds[i], mIconUnselectIds[i]))
        }
        tab_layout!!.setTabData(mTabEntities)
        //点击监听
        tab_layout!!.setOnTabSelectListener(object : OnTabSelectListener {
            override fun onTabSelect(position: Int) {
                SwitchTo(position)
            }

            override fun onTabReselect(position: Int) {}
        })
    }

    /**
     * 初始化碎片
     */
    private fun initFragment(savedInstanceState: Bundle?) {
        val transaction = supportFragmentManager.beginTransaction()
        var currentTabPosition = 0
        if (savedInstanceState != null) {
            mFirstFragment = supportFragmentManager.findFragmentByTag("newsMainFragment") as FirstFragment?
            mSecondFragment = supportFragmentManager.findFragmentByTag("photosMainFragment") as SecondFragment?
            mThirdFragment = supportFragmentManager.findFragmentByTag("videoMainFragment") as ThirdFragment?
            mFourFragment = supportFragmentManager.findFragmentByTag("careMainFragment") as FourFragment?
            currentTabPosition = savedInstanceState.getInt(AppConstant.HOME_CURRENT_TAB_POSITION)
        } else {
            mFirstFragment = FirstFragment()
            mSecondFragment = SecondFragment()
            mThirdFragment = ThirdFragment()
            mFourFragment = FourFragment()

            transaction.add(R.id.fl_body, mFirstFragment!!, "newsMainFragment")
            transaction.add(R.id.fl_body, mSecondFragment!!, "photosMainFragment")
            transaction.add(R.id.fl_body, mThirdFragment!!, "videoMainFragment")
            transaction.add(R.id.fl_body, mFourFragment!!, "careMainFragment")
        }
        transaction.commit()
        SwitchTo(currentTabPosition)
        tab_layout!!.currentTab = currentTabPosition
    }

    /**
     * 切换
     */
    private fun SwitchTo(position: Int) {
        LogUtils.logd("主页菜单position$position")
        val transaction = supportFragmentManager.beginTransaction()
        when (position) {
            0 -> {
                transaction.hide(mSecondFragment!!)
                transaction.hide(mThirdFragment!!)
                transaction.hide(mFourFragment!!)
                transaction.show(mFirstFragment!!)
                transaction.commitAllowingStateLoss()
            }
            1 -> {
                transaction.hide(mFirstFragment!!)
                transaction.hide(mThirdFragment!!)
                transaction.hide(mFourFragment!!)
                transaction.show(mSecondFragment!!)
                transaction.commitAllowingStateLoss()
            }
            2 -> {
                transaction.hide(mFirstFragment!!)
                transaction.hide(mSecondFragment!!)
                transaction.hide(mFourFragment!!)
                transaction.show(mThirdFragment!!)
                transaction.commitAllowingStateLoss()
            }
            3 -> {
                transaction.hide(mFirstFragment!!)
                transaction.hide(mSecondFragment!!)
                transaction.hide(mThirdFragment!!)
                transaction.show(mFourFragment!!)
                transaction.commitAllowingStateLoss()
            }
            else -> {
            }
        }
    }

    /**
     * 菜单显示隐藏动画
     *
     * @param showOrHide
     */
    private fun startAnimation(showOrHide: Boolean) {
        val layoutParams = tab_layout!!.layoutParams
        val valueAnimator: ValueAnimator
        val alpha: ObjectAnimator
        if (!showOrHide) {
            valueAnimator = ValueAnimator.ofInt(tabLayoutHeight, 0)
            alpha = ObjectAnimator.ofFloat(tab_layout, "alpha", 1f, 0f)
        } else {
            valueAnimator = ValueAnimator.ofInt(0, tabLayoutHeight)
            alpha = ObjectAnimator.ofFloat(tab_layout, "alpha", 0f, 1f)
        }
        valueAnimator.addUpdateListener {
            layoutParams.height = valueAnimator.animatedValue as Int
            tab_layout!!.layoutParams = layoutParams
        }
        val animatorSet = AnimatorSet()
        animatorSet.duration = 500
        animatorSet.playTogether(valueAnimator, alpha)
        animatorSet.start()
    }

    /**
     * 监听全屏视频时返回键
     */
    override fun onBackPressed() {
        if (System.currentTimeMillis() - exitTime > 2000) {
            Toast.makeText(applicationContext, "再按一次退出程序", Toast.LENGTH_SHORT).show()
            exitTime = System.currentTimeMillis()
        } else {
            val backHome = Intent(Intent.ACTION_MAIN)
            backHome.flags = Intent.FLAG_ACTIVITY_NEW_TASK
            backHome.addCategory(Intent.CATEGORY_HOME)
            startActivity(backHome)
        }
    }


    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        //崩溃前保存位置
        if (tab_layout != null) {
            outState.putInt(AppConstant.HOME_CURRENT_TAB_POSITION, tab_layout!!.currentTab)
        }
    }


    override fun onDestroy() {
        super.onDestroy()
        ChangeModeController.onDestory()
    }

    companion object {
        private var tabLayoutHeight: Int = 0

        /**
         * 入口
         *
         * @param activity
         */
        fun startAction(activity: Activity) {
            val intent = Intent(activity, MainActivity::class.java)
            activity.startActivity(intent)
            activity.overridePendingTransition(R.anim.fade_in,
                    com.jaydenxiao.common.R.anim.fade_out)
        }
    }

}
